spoon-smpl/src/test/java/spoon/smpl/MetavarsTest.java

```diff
@@ -0,0 +1,145 @@
+package spoon.smpl;
+
+import fr.inria.controlflow.ControlFlowNode;
+import org.junit.Before;
+import org.junit.Test;
+import spoon.smpl.formula.*;
+import spoon.smpl.metavars.ConstantConstraint;
+import spoon.smpl.metavars.ExpressionConstraint;
+import spoon.smpl.metavars.IdentifierConstraint;
+import spoon.smpl.metavars.TypeConstraint;
+
+import java.lang.reflect.Field;
+import java.util.*;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+import static spoon.smpl.TestUtils.*;
+import static spoon.smpl.TestUtils.parseReturnStatement;
+
+/**
+ * This is essentially an integration test of most of the SmPL stack, leaving out the parser.
+ */
+public class MetavarsTest {
+    @Before
+    public void resetControlFlowNodeCounter() {
+        // needed for consistent IDs in CFGModels
+        try {
+            Field field = ControlFlowNode.class.getDeclaredField("count");
+            field.setAccessible(true);
+            ControlFlowNode.count = 0;
+        } catch (Exception e) {
+            fail("Unable to reset ControlFlowNode id counter");
+        }
+    }
+
+    private static Formula stmt(String code, Map<String, ParameterPostProcessStrategy> metavars) {
+        return new StatementPattern(makePattern(parseStatement(code), new ArrayList<>(metavars.keySet())), metavars);
+    }
+
+    private static Formula retstmt(String code, Map<String, ParameterPostProcessStrategy> metavars) {
+        return new StatementPattern(makePattern(parseReturnStatement(code), new ArrayList<>(metavars.keySet())), metavars);
+    }
+
+    @Test
+    public void testCompatibleBindings() {
+
+        // contract: metavariables bound to "same thing" in different nodes are joined under AND
+        // TODO: add tests for other formula connectors such as OR
+
+        Model modelA = new CFGModel(methodCfg(parseMethod("int m() { int x = 1; return x; }")));
+        ModelChecker checkerA = new ModelChecker(modelA);
+        //System.out.println(((CFGModel) modelA).getCfg().toGraphVisText());
+
+        Map<String, ParameterPostProcessStrategy> meta = metavars("z", new IdentifierConstraint());
+
+        stmt("int z = 1;", meta).accept(checkerA);
+        assertEquals("[(4, {z=x})]", checkerA.getResult().toString());
```

Comment:
what does the "4" mean here?

Reply:
It's the ID of a state that satisfied the formula. For a CFGModel, each node in the CFG corresponds to a state and the state IDs are determined by the node IDs (`ControlFlowNode::getId`). A handy way to get a picture of the node IDs is to take the GraphViz output available from `ControlFlowGraph::toGraphVisText` and rendering it: [example](https://dreampuf.github.io/GraphvizOnline/#digraph%20null%20%7B%20%0Anode%20%5Bfontsize%20%3D%208%5D%3B%0A1%20%5Bshape%3DMdiamond%2C%20label%3D%22BEGIN_2%20%22%5D%3B%0A2%20%5Bshape%3Drectangle%2C%20label%3D%224%20-%20int%20x%20%3D%201%20%22%5D%3B%0A3%20%5Bshape%3Drectangle%2C%20label%3D%225%20-%20return%20x%20%22%5D%3B%0A4%20%5Bshape%3Ddoublecircle%2C%20label%3D%22EXIT_1%20%22%5D%3B%0A2%20-%3E%203%20%3B%0A%203%20-%3E%204%20%3B%0A%201%20-%3E%202%20%3B%0A%20%0A%20%7D%0A%0AProcess%20finished%20with%20exit%20code%200%0A)

Reply:
very interesting. could you add this explanation as comment in the test?